### ISSUE TITLE
ZD: chefspec error with ChefWorkstation 20.6.62 but works with Chef Development Kit

### ISSUE THREAD

**Issue by dheerajd-msys** [chef/chef-workstation#1459]
## Description

When running chefspec, the following error occurs.
```
$ chef -v
Chef Workstation version: 20.6.62
Chef Infra Client version: 16.1.16
Chef InSpec version: 4.19.0
Chef CLI version: 3.0.4
Test Kitchen version: 2.5.1
Cookstyle version: 6.7.3

$ chef exec rspec spec/unit/recipes/batch_spec.rb
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chefspec-9.1.0/lib/chefspec/extensions/chef/resource.rb:139: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-16.1.16/lib/chef/resource.rb:1343: warning: The called method `provides' is defined here

// １回目のconverge。２回目のconvergeがありません。そのためnode[fqdn]に値が入らず、specが失敗しているのではないかと予想しています。
debug print start

fqdn= fauxhai.local
'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"]
idx= 0
debug print end

F

Failures:

1) test_yj_chef_tutorial_cookbook::batch moritoki-test set cron at minute:0
Failure/Error: expect(chef_run).to create_cron('moritoki-test').with(minute: '0', command: '/usr/bin/test')

expected "cron[moritoki-test]" to have parameters:

minute "0", was 0
# ./spec/unit/recipes/batch_spec.rb:18:in `block (3 levels) in <top (required)>'

Finished in 1.78 seconds (files took 6.16 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/unit/recipes/batch_spec.rb:15 # test_yj_chef_tutorial_cookbook::batch moritoki-test set cron at minute:0
```
In ChefDK instead, the same cookbook does not have any error.
```
$ chef -v
Chef Development Kit Version: 3.13.1
chef-client version: 14.14.29
delivery version: master (4b21ec7e07fdfa82e86aa80e4f2372dde8e368bb)
berks version: 7.0.8
kitchen version: 1.25.0
inspec version: 3.9.3

$ chef exec rspec spec/unit/recipes/batch_spec.rb
Fetched 'redhat/6.10' from GitHub, but could not write to the local path: /opt/chefdk/embedded/lib/ruby/gems/2.5.0/gems/fauxhai-6.11.0/lib/fauxhai/platforms/redhat/6.10.json. Fix the local file permissions to avoid downloading this file every run.

// １回目、自分の想定では、convergeの回数だけ、batch_spec.rbが読み込まれるイメージでした。
debug print start

fqdn= fauxhai.local　　 // 最初のconvergeではセットしてないので問題ない
'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"]
idx= 0
debug print end

// ２回目
debug print start

fqdn= t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp // セットされた
'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"]
idx= 0
debug print end

.

Finished in 0.6958 seconds (files took 2.21 seconds to load)
1 example, 0 failures
```
## Chef Workstation Version

Chef Workstation version: 20.6.62
## Related Issue
https://github.com/chef/customer-bugs/issues/240

**NOTE:** Please check the cookbook attachment in associated issue

**Comment by tyler-ball** [comment]
Thanks @dheerajd-msys - looks like the issue is the difference between expecting a string and receiving an integer (`minute "0", was 0`). We'll have to look and determine whether ChefDK was incorrectly having this test succeed. In any case, thanks for bringing this to our attention!

**Comment by Nirbhay1997** [comment]
@tyler-ball 
As the issue is replicable these are some of the Observations :

With ChefDK  latest  
When :  (minute: '0')(with quotes)

Result: Pass
Output :
`debug print start fqdn= fauxhai.local 'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"] idx= 0 debug print end . Finished in 0.14431 seconds (files took 1.94 seconds to load) 1 example, 0 failures`

When :  (minute: 0)(Without quotes)
Result: Fail
Output :
` debug print start fqdn= fauxhai.local 'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"] idx= 0 debug print end F Failures: 1) test_yj_chef_tutorial_cookbook::batch moritoki-test set cron at minute:0 Failure/Error: expect(chef_run).to create_cron('moritoki-test').with(minute: 0, command: '/usr/bin/test') expected "cron[moritoki-test]" to have parameters: minute 0, was "0" # ./spec/unit/recipes/batch_spec.rb:20:in `block (3 levels) in <top (required)>' Finished in 0.13739 seconds (files took 1.99 seconds to load) 1 example, 1 failure`



With Chef workstation: 20.6.62
When :  (minute: '0')
Result : Fail 
Output :

`opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chefspec-9.1.0/lib/chefspec/extensions/chef/resource.rb:139: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call /opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-16.2.44/lib/chef/resource.rb:1354: warning: The called method `provides' is defined here debug print start fqdn= fauxhai.local 'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"] idx= 0 debug print end F Failures: 1) test_yj_chef_tutorial_cookbook::batch moritoki-test set cron at minute:0 Failure/Error: expect(chef_run).to create_cron('moritoki-test').with(minute: "0", command: '/usr/bin/test') expected "cron[moritoki-test]" to have parameters: minute "0", was 0 # ./spec/unit/recipes/batch_spec.rb:20:in `block (3 levels) in <top (required)>' Finished in 0.13506 seconds (files took 3.36 seconds to load) 1 example, 1 failure Failed examples: rspec ./spec/unit/recipes/batch_spec.rb:16 # test_yj_chef_tutorial_cookbook::batch moritoki-test set cron at minute:0
`

When :  (minute: 0)(Without quotes)
Result : Fail 
Output :
`/opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chefspec-9.1.0/lib/chefspec/extensions/chef/resource.rb:139: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call /opt/chef-workstation/embedded/lib/ruby/gems/2.7.0/gems/chef-16.2.44/lib/chef/resource.rb:1354: warning: The called method `provides' is defined here debug print start fqdn= fauxhai.local 'node['collector']['batch_hosts']'=["t-tsumoritoki01.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki02.chef.ssk.ynwm.yahoo.co.jp", "t-tsumoritoki03.chef.ssk.ynwm.yahoo.co.jp"] idx= 0 debug print end . Finished in 0.1366 seconds (files took 3.23 seconds to load) 1 example, 0 failures
`

**Comment by clintoncwolfe** [comment]
I believe I have traced the issue to a [change](https://github.com/chef/chef/pull/9522) introduced in Chef Infra 16, which improves the validations performed on the properties of the `cron` resource. Prior to this change, the cron resource always stringified the property, but now the resource does not stringify the property until it actually writes the crontab.

As this is a change at a major version number of Chef Infra, it is a breaking change.  However, I suggest that you have a couple of paths forward. 

One option is to move to Chef Infra 16, and change your ChefSpec test to expect an integer.

Another option, if you must have your recipe and test pass in both Chef Infra 16 and older versions, is to modify the recipe - not the test - to stringify the minute property:

```
cron_minute = (idx * 20).to_s
```

I was able to test this under an older Chef Infra (version 15.10.12, from Chef Workstation 0.18.3, which pre-dates the validation changes), as well as under Chef Infra 16.4.41.
